src/calendar/google/mod.rs

```diff
@@ -0,0 +1,364 @@
+    #[error("Authentication error: {0}")]
+    GCloudAuthError(#[from] google_cloud_auth::error::Error),
+
+    /// Error while making a http request
+    #[error("RequestError: {0}")]
+    RequestError(#[from] reqwest::Error),
+
+    /// Error while executing some middleware code
+    #[error("RequestMiddlewareError: {0}")]
+    RequestMiddlewareError(#[from] reqwest_middleware::Error),
+
+    /// Error while building http headers
+    #[error("InvalidHeaderError: {0}")]
+    RequestInvalidHedaer(#[from] reqwest::header::InvalidHeaderValue),
+
+    /// No calendar id is found in env when it is not passed into client constructor
+    #[error("MissingCalendarIDError: {0}")]
+    MissingCalendarIDError(#[from] std::env::VarError),
+
+    /// Error while parsing responses
+    #[error("JsonParsingError: {0}")]
+    JsonParsingError(#[from] serde_json::Error),
+}
+
+async fn refresh_token() -> Result<google_cloud_auth::token::Token, ClientError> {
+    let scopes = ["https://www.googleapis.com/auth/calendar.readonly"];
+    let config = google_cloud_auth::Config {
+        audience: None,
+        scopes: Some(&scopes),
+    };
+
+    // This internally looks up the service account credentials that come from json key file
+    // generated in the google cloud console. The lookup happens via the
+    // GOOGLE_APPLICATION_CREDENTIALS environments stored in the .env file
+    let ts = google_cloud_auth::create_token_source(config).await?;
+    Ok(ts.token().await?)
+}
+
+impl From<ClientError> for reqwest_middleware::Error {
+    fn from(err: ClientError) -> Self {
+        reqwest_middleware::Error::Middleware(anyhow::Error::new(err))
+    }
+}
+
+struct AuthMiddleware {
+    token: Mutex<Option<google_cloud_auth::token::Token>>,
+}
+
+impl AuthMiddleware {
+    fn new() -> Self {
+        Self {
+            token: Mutex::new(None),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Middleware for AuthMiddleware {
+    async fn handle(
+        &self,
+        req: Request,
+        extensions: &mut Extensions,
+        next: Next<'_>,
+    ) -> reqwest_middleware::Result<Response> {
+        let mut token = self.token.lock().await;
+        if !(token.is_some() && token.as_ref().unwrap().valid()) {
+            match refresh_token().await {
+                Ok(tok) => *token = Some(tok),
+                Err(e) => return Err(reqwest_middleware::Error::from(e)),
+            };
+        }
+        // should be safe to call unwrap here if refresh_token() fails the error is caught above so
+        // we should always have a valid token here
+        let access_token = &token.as_ref().unwrap().access_token;
+
+        // insert auth header
+        let mut req = req;
+        req.headers_mut().insert(
+            header::AUTHORIZATION,
+            header::HeaderValue::from_str(format!("Bearer {}", access_token).as_str())
+                .map_err(|err| ClientError::from(err))?,
+        );
+
+        next.run(req, extensions).await
+    }
+}
```

Comment:
I have a different suggestion for implementing the middleware. Here are the key improvements:

- use `match` to check if a token exists and is valid. this also avoids having the comment that states why unwrapping the token is safe.
- replace error handling of `refresh_token` with just `?`
- mark the header value as sensitive
- add a scope so that the mutex is unlocked before passing the request to the next handler

This also fixes a clippy warning.

```suggestion
impl Middleware for AuthMiddleware {
    async fn handle(
        &self,
        mut req: Request,
        extensions: &mut Extensions,
        next: Next<'_>,
    ) -> reqwest_middleware::Result<Response> {
        // Create a new scope to ensure that the mutex is unlocked before we call the next
        // middleware. We don't need to lock longer than needed.
        {
            let mut token = self.token.lock().await;

            let token = match token.as_ref() {
                Some(token) if token.valid() => token,
                // We don't have a token yet, or the current one expired, let's refresh it.
                _ => refresh_token().await.map(|t| {
                    *token = Some(t);
                    token.as_ref().unwrap()
                })?,
            };

            let mut header =
                header::HeaderValue::from_str(&format!("Bearer {}", token.access_token))
                    .map_err(ClientError::from)?;
            header.set_sensitive(true);

            req.headers_mut().insert(header::AUTHORIZATION, header);
        }

        next.run(req, extensions).await
    }
}
```